### DISCUSSION TITLE
porting Lua to ELKS?

### DISCUSSION THREAD

**Discussion by sergione78** [jbruchon/elks#1404] [General]
I was wondering if there is the possibility of porting the Lua language to Elks. I tried to use the ia16-elf-gcc compiler by myself editing the Makefile, but unfortunately I have absolutely no experience with compiling and everything ends up with a file not found. Would anyone better than me want to try?

**Comment by ghaerr** [comment]
Hello @sergione78,

Before editing Makefiles and attempting a port, it is important to take a look at the Lua engine itself, and try to learn whether or not it is dependent on the size of integers or addresses being 32-bits (or greater). This can also usually be quickly learned by looking at memory allocations - ELKS is limited to a max single allocation of 32767 bytes, and a further max combined allocation of 64K bytes. By most software standards of today, this is extremely small.

Another place to look would be the Lua "marketing" literature (so to speak) - does it say "Can run on 16-bit microcontrollers"? Without a lot of work, ELKS applications are limited to 64K data, and 128K code.

I hope this helps.

Thank you!

**Reply by sergione78** [reply]
Thank you for your answer. Actually, I made a little progress yesterday, but the ia16 compiler was giving me an error about the block size. Maybe 128kb for code isn't enough. The DOS executable is 150kb more or less, maybe there is a way to narrow it down (compilation optimizations, I don't know). As for the first problem you mentioned, there is a configuration file that can be managed to have 16-bit data types, and I have already modified this. In my opinion it might be doable, but I'm too newbie to do it easily.

**Reply by sergione78** [reply]
just needs ANSI C capable compiler (C89 compliant).

**Reply by sergione78** [reply]
a further little progress. Now it does compile the sources, but it's giving a lot of missing refecences in the link section. 
Here is the console output:

make ansi
cd src && make ansi
make[1]: Entering directory '/home/lineageos/Desktop/Lua_5_1/lua-5.1.5/src'
make all MYCFLAGS=-DLUA_ANSI
make[2]: Entering directory '/home/lineageos/Desktop/Lua_5_1/lua-5.1.5/src'
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lapi.o lapi.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lcode.o lcode.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o ldebug.o ldebug.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o ldo.o ldo.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o ldump.o ldump.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lfunc.o lfunc.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lgc.o lgc.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o llex.o llex.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lmem.o lmem.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lobject.o lobject.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lopcodes.o lopcodes.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lparser.o lparser.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lstate.o lstate.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lstring.o lstring.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o ltable.o ltable.c
In file included from ltable.c:35:0:
ltable.c: In function ‘mainposition’:
llimits.h:42:37: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
 #define IntPoint(p)  ((unsigned int)(lu_mem)(p))
                                     ^
ltable.h:13:32: note: in definition of macro ‘gnode’
 #define gnode(t,i) (&(t)->node[i])
                                ^
ltable.c:63:26: note: in expansion of macro ‘hashmod’
 #define hashpointer(t,p) hashmod(t, IntPoint(p))
                          ^~~~~~~
ltable.c:63:37: note: in expansion of macro ‘IntPoint’
 #define hashpointer(t,p) hashmod(t, IntPoint(p))
                                     ^~~~~~~~
ltable.c:109:14: note: in expansion of macro ‘hashpointer’
       return hashpointer(t, pvalue(key));
              ^~~~~~~~~~~
llimits.h:42:37: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
 #define IntPoint(p)  ((unsigned int)(lu_mem)(p))
                                     ^
ltable.h:13:32: note: in definition of macro ‘gnode’
 #define gnode(t,i) (&(t)->node[i])
                                ^
ltable.c:63:26: note: in expansion of macro ‘hashmod’
 #define hashpointer(t,p) hashmod(t, IntPoint(p))
                          ^~~~~~~
ltable.c:63:37: note: in expansion of macro ‘IntPoint’
 #define hashpointer(t,p) hashmod(t, IntPoint(p))
                                     ^~~~~~~~
ltable.c:111:14: note: in expansion of macro ‘hashpointer’
       return hashpointer(t, gcvalue(key));
              ^~~~~~~~~~~
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o ltm.o ltm.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lundump.o lundump.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lvm.o lvm.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lzio.o lzio.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lauxlib.o lauxlib.c
lauxlib.c: In function ‘luaL_loadfile’:
lauxlib.c:577:4: warning: this ‘while’ clause does not guard... [-Wmisleading-indentation]
    while ((c = getc(lf.f)) != EOF && c != LUA_SIGNATURE[0]) ;
    ^~~~~
lauxlib.c:578:5: note: ...this statement, but the latter is misleadingly indented as if it is guarded by the ‘while’
     lf.extraline = 0;
     ^~
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lbaselib.o lbaselib.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o ldblib.o ldblib.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o liolib.o liolib.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lmathlib.o lmathlib.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o loslib.o loslib.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o ltablib.o ltablib.c
ltablib.c: In function ‘addfield’:
ltablib.c:137:3: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
   if (!lua_isstring(L, -1))
   ^~
ltablib.c:140:5: note: ...this statement, but the latter is misleadingly indented as if it is guarded by the ‘if’
     luaL_addvalue(b);
     ^~~~~~~~~~~~~
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lstrlib.o lstrlib.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o loadlib.o loadlib.c
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o linit.o linit.c
ia16-elf-gcc-ar rcu liblua.a lapi.o lcode.o ldebug.o ldo.o ldump.o lfunc.o lgc.o llex.o lmem.o lobject.o lopcodes.o lparser.o lstate.o lstring.o ltable.o ltm.o lundump.o lvm.o lzio.o lauxlib.o lbaselib.o ldblib.o liolib.o lmathlib.o loslib.o ltablib.o lstrlib.o loadlib.o linit.o	# DLL needs all object files
ranlib liblua.a
ia16-elf-gcc -Os -Wall -DLUA_ANSI   -c -o lua.o lua.c
ia16-elf-gcc -o lua  lua.o liblua.a -lm -melks-libc -ffreestanding
/usr/lib/x86_64-linux-gnu/gcc/ia16-elf/6.3.0/../../../../../ia16-elf/bin/ld.gold: error: /usr/lib/x86_64-linux-gnu/gcc/ia16-elf/6.3.0/../../../../../ia16-elf/lib/elkslibc/libc.a(tp-main.o): multiple definition of 'main'
/usr/lib/x86_64-linux-gnu/gcc/ia16-elf/6.3.0/../../../../../ia16-elf/bin/ld.gold: lua.o: previous definition here
/usr/lib/x86_64-linux-gnu/gcc/ia16-elf/6.3.0/../../../../../ia16-elf/bin/ld.gold: error: address of section '.fartext' moves backward from 0x33f50 to 0x20000
/usr/lib/x86_64-linux-gnu/gcc/ia16-elf/6.3.0/../../../../../ia16-elf/bin/ld.gold: error: load segment overlap [0x10000 -> 0x33f50] and [0x30000 -> 0x33520]
lua.o:function l_message: error: undefined reference to '_impure_ptr'
lua.o:function l_message: error: undefined reference to '_impure_ptr'
lua.o:function l_message: error: undefined reference to '_impure_ptr'
lua.o:function pushline: error: undefined reference to '_impure_ptr'
liblua.a(ldo.o):function luaD_rawrunprotected: error: undefined reference to 'setjmp'
liblua.a(ldo.o):function luaD_throw: error: undefined reference to 'longjmp'
liblua.a(lobject.o):function luaO_str2d: error: undefined reference to '__ctype_ptr__'
liblua.a(lvm.o):function l_strcmp: error: undefined reference to 'strcoll'
liblua.a(lauxlib.o):function __sgetc_r: error: undefined reference to '__srget_r'
liblua.a(lauxlib.o):function __sgetc_r: error: undefined reference to '__srget_r'
liblua.a(lauxlib.o):function errfile: error: undefined reference to '__errno'
liblua.a(lauxlib.o):function getF: error: undefined reference to 'feof'
liblua.a(lauxlib.o):function luaL_loadfile: error: undefined reference to 'ferror'
liblua.a(llex.o):function luaX_token2str: error: undefined reference to '__ctype_ptr__'
liblua.a(llex.o):function read_numeral: error: undefined reference to '__ctype_ptr__'
liblua.a(llex.o):function read_numeral: error: undefined reference to '__ctype_ptr__'
liblua.a(llex.o):function read_numeral: error: undefined reference to 'localeconv'
liblua.a(liolib.o):function pushresult: error: undefined reference to '__errno'
liblua.a(liolib.o):function g_read: error: undefined reference to 'clearerr'
liblua.a(liolib.o):function g_read: error: undefined reference to 'ferror'
liblua.a(liolib.o):function g_read: error: undefined reference to '__srget_r'
liblua.a(liolib.o):function g_read: error: undefined reference to '__srget_r'
liblua.a(liolib.o):function io_readline: error: undefined reference to 'ferror'
liblua.a(liolib.o):function io_readline: error: undefined reference to '__errno'
liblua.a(liolib.o):function io_tmpfile: error: undefined reference to 'tmpfile'
liblua.a(liolib.o):function fileerror.constprop.8: error: undefined reference to '__errno'
liblua.a(loslib.o):function os_setlocale: error: undefined reference to 'setlocale'
liblua.a(loslib.o):function os_remove: error: undefined reference to 'remove'
liblua.a(loslib.o):function os_difftime: error: undefined reference to 'difftime'
liblua.a(loslib.o):function os_clock: error: undefined reference to 'clock'
liblua.a(loslib.o):function os_date: error: undefined reference to 'strftime'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:55: lua] Error 1
make[2]: Leaving directory '/home/lineageos/Desktop/Lua_5_1/lua-5.1.5/src'
make[1]: *** [Makefile:87: ansi] Error 2
make[1]: Leaving directory '/home/lineageos/Desktop/Lua_5_1/lua-5.1.5/src'
make: *** [Makefile:56: ansi] Error 2

**Comment by ghaerr** [comment]
Hello @sergione78,

Thank you for the logs. To be frank, the likelihood of success on this is not looking good:

> llimits.h:42:37: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
#define IntPoint(p) ((unsigned int)(lu_mem)(p))

This error is basically saying that sizeof(lu_mem) != sizeof unsigned int). On ia16-elf-gcc, sizeof(int) is always 16, so sizeof(lu_mem) is probably 32, which won't work on ELKS. You'll want to look further into the definition of `lu_mem`, and see how it might be made into a 16-bit pointer.

> /usr/lib/x86_64-linux-gnu/gcc/ia16-elf/6.3.0/../../../../../ia16-elf/bin/ld.gold: error: address of section '.fartext' moves backward from 0x33f50 to 0x20000
/usr/lib/x86_64-linux-gnu/gcc/ia16-elf/6.3.0/../../../../../ia16-elf/bin/ld.gold: error: load segment overlap [0x10000 -> 0x33f50] and [0x30000 -> 0x33520]

This error means that the code segment is too large, and that you'll have to use medium model (which could also easily be too small). This can be done using ia16-elf-gcc using the following flags:
```
CFLBASE = -fno-inline -melks-libc -mtune=i8086 -Wall -Os
CFLBASE += -mcmodel=medium -msegment-relocation-stuff -ffunction-sections
CFLAGS  = $(CFLBASE) $(WARNINGS) $(LOCALFLAGS) $(INCLUDES)
.c.o:
    $(CC) $(CFLAGS) -c -o $*.o $<
```
There are also some link problems which have to be ironed out by adding/subtracting .o files, regarding the undefined or multiply defined symbols.

Thank you!

**Reply by sergione78** [reply]
I've been trying for at least three days to get it to compile (also using the parameters you provided), but I always get errors. These errors, however, are IMHO due to problems in the toolchain and not the fact that Lua requires 32bit machines. In fact, the language has been successfully compiled in the Atari TOS environment, and there is also a version for MS-DOS (I don't know if DPMI). In any case, the configuration files expressly provide for the possibility of using 16-bit machines as shown in the fragment below


/*
** $Id: llimits.h $
** Limits, basic types, and some other 'installation-dependent' definitions
** See Copyright Notice in lua.h
*/

#ifndef llimits_h
#define llimits_h


#include <limits.h>
#include <stddef.h>


#include "lua.h"


/*
** 'lu_mem' and 'l_mem' are unsigned/signed integers big enough to count
** the total memory used by Lua (in bytes). Usually, 'size_t' and
** 'ptrdiff_t' should work, **but we use 'long' for 16-bit machines**.
*/
#if defined(LUAI_MEM)		/* { external definitions? */
typedef LUAI_UMEM lu_mem;
typedef LUAI_MEM l_mem;
#elif LUAI_IS32INT	/* }{ */
typedef size_t lu_mem;
typedef ptrdiff_t l_mem;
#else  /* 16-bit ints */	/* }{ */
typedef unsigned long lu_mem;
typedef long l_mem;
#endif				/* } */

**Reply by sergione78** [reply]
I'm sorry if I can't be more precise, but as you have certainly understood I'm not really an expert

**Comment by ghaerr** [comment]
Hello @sergione78,

From the looks of the comment in llimits.h, lua may in fact be capable of running on 16-bit machines, so that is hopeful.

With regards to the continuing errors, I need to see the exact error messages from the toolchain in order to help you. I doubt we're looking at any toolchain errors, but from your previous posts, it looks like there may be a number of undefined C library functions, (like feof(), clearerr(), which are macros and require including stdio.h, and setjmp(), which requires setjmp.h, etc). There are also some undefined calls to functions like strftime and ___errno, which are not (yet) in ELKS' C library. I also noticed a duplicate `main` which means two applications were attempting to be linked.  It is also possible that the code size is > 128K, which we can't work around, unless Lua can be configured to build with fewer features.

I'd say you're doing pretty good actually. Please post more linker output and we can discuss from there.

Thank you!

**Comment by rafael2k** [comment]
I missed this discussion. I tried porting Lua with ia16-gcc, but the text will not fit unless large memory is used.
I ported Lua 5.1.5 using OpenWatcom:
https://github.com/rafael2k/lua

**Comment by rafael2k** [comment]
https://github.com/ghaerr/elks/discussions/2234

**Comment by toncho11** [comment]
You can try it here in this ELKS bootable hdd image:
https://github.com/ghaerr/elks/discussions/2240
You can boot it in: https://copy.sh/v86/

**Comment by sergione78** [comment]
I'm really glad you managed to port Lua to ELKS. As I imagined, changing toolchain and using OWC and large model did the trick. Bravo!

**Comment by toncho11** [comment]
We have a lua script that rotates a cube (not just an image of a cube) and the cube is dispalyed using Nano-X API:

<img width="640" height="480" alt="image" src="https://github.com/user-attachments/assets/9160014e-d638-49f5-bd1b-e084952c2fb8" />

**Reply by rafael2k** [reply]
This is really cool @toncho11 !! 
We could think about changing the default Lua to 5.1.

**Comment by toncho11** [comment]
We also have pong.lua now:

<img width="640" height="480" alt="image" src="https://github.com/user-attachments/assets/ac154f6c-8e11-400f-be4b-f2dd17bc96bc" />

Keyboard handling is missing right now, so it works in ai vs ai mode right now.

In copy.86 it works well, but on a 286 emulator is very slow, but could be optimized to some degree.

**Reply by rafael2k** [reply]
This is pretty cool! We could try to use the integer-only Lua port (and use integer only math only).

**Reply by rafael2k** [reply]
this branch: https://github.com/rafael2k/lua/commits/lua-5.1i

**Reply by toncho11** [reply]
I am not sure the math is the main problem. Here https://github.com/rafael2k/lua/issues/13 I propose a few optimisations that can be applied to lua 5.1/5.5.
